### PR TITLE
add Thread/get support & References to previous method results

### PR DESCRIPTION
**Pull request by qyb**

### CODE REVIEW

**Comment by alagane** [comment]
Hello, isn't the method thread_get part missing from the index.ts file?

**Comment by qyb** [comment]
> Hello, isn't the method thread_get part missing from the index.ts file?

I'm trying to add thread test with james:memory. I want to implement one request combining 4 methods (Email/query + Email/get + Thread/get + Email/get) as the description in https://jmap.io/client.html#cold-boot

But I cannot get thread work. I forged some .eml files with related Message-ID/In-Reply-To/References header. Then I use james-cli ImportEml command to import them. But the system didn't recognize threads. May you give me some tips?

**Comment by qyb** [comment]
> > But I cannot get thread work. I forged some .eml files with related Message-ID/In-Reply-To/References header. Then I use james-cli ImportEml command to import them. But the system didn't recognize threads. May you give me some tips?
> 
> @quantranhong1999 ?

I have done my research about memory-app ThreadIdGuessingAlgorithm implementation. Maybe it was just a naiveAlgorighm?

Anyway, it won't stop the test's logical.:)

**Comment by qyb** [comment]
> > I have done my research about memory-app ThreadIdGuessingAlgorithm implementation. Maybe it was just a naiveAlgorighm?
> 
> We are using SearchThreadIdGuessingAlgorithm (search on memory) implementation for memory-app.

docker exec e26fd3fdb97c james-cli AddUser user1@localhost password

provisioning mailbox (by fetchSession/login...)


docker exec e26fd3fdb97c james-cli ListUserMailboxes user1@localhost
docker exec e26fd3fdb97c james-cli ImportEml "#private" user1@localhost INBOX /root/conf/1.eml
docker exec e26fd3fdb97c james-cli ImportEml "#private" user1@localhost Sent /root/conf/2.eml
docker exec e26fd3fdb97c james-cli ImportEml "#private" user1@localhost INBOX /root/conf/3.eml

I expect INBOX should be only 1 thread. but mailbox_get shows two threads in INBOX
[eml.zip](https://github.com/linagora/jmap-client-ts/files/8924278/eml.zip)
.

**Comment by quantranhong1999** [comment]
What is the James memory version you are using? BTW can you try Email/query and then Email/get threadId properties to see the mapping between messageId and threadId first?

**Comment by quantranhong1999** [comment]
Good to hear that :')

**Comment by chibenwa** [comment]
> But collapseThreads result is still multiple threads.

We (in James) did not implement it just yet (as it would require to push aggregation down to the search engine)

Contributions welcomed for the willing ;-)

**Comment by qyb** [comment]
> > But collapseThreads result is still multiple threads.
> 
> We (in James) did not implement it just yet (as it would require to push aggregation down to the search engine)
> 
> Contributions welcomed for the willing ;-)

Now, this pull request just remains `ResultReference` test. :)

**Review by chibenwa — DISMISSED** [review]

**Comment by qyb** [comment]
> How is the keystore generated? I used a keypair on my pull request #62 I don't know which method is better, but I personally prefair to not have binary files is possible. I think the best method would be to generate the keypair/keystore when running tests if none is found on the directory, and ignore it on .gitignore

https://www.npmjs.com/package/keypair is our need.

I'll remove ResultReference object from this pr because now I think it should be in another pull request. This PR is too complex and putting multiple arguments in one request is another complex thing.

File: src/types.ts [hunk 1]
```diff
@@ -118,6 +134,7 @@ export interface IQueryArguments<FilterCondition extends IFilterCondition>
   anchorOffset?: number;
   limit?: number;
   calculateTotal?: boolean;
+  collapseThreads?: boolean;
```
[the remark this answers is not in the document]
Comment (qyb) [reply src/types.ts:137]: I'll remove collapseThreads first. Let's do it in another pull request.

I have learned wide knowledge about jest/typescript from this PR.

Thanks to all of you.

**Review by alagane — APPROVED** [review]

**Comment by alagane** [comment]
Ready to merge for me.
Thank you for your contribution!

**Pull request merged**